src/tamp_improv/approaches/improvisational/policies/rl2mpc.py

```diff
@@ -0,0 +1,196 @@
+    RLConfig,
+    RLPolicy,
+    TrainingProgressCallback,
+)
+
+
+@dataclass
+class RL2MPCConfig:
+    """Configuration for RL2MPC policy."""
+
+    rl_config: RLConfig = field(default_factory=RLConfig)
+    mpc_config: MPCConfig = field(default_factory=MPCConfig)
+    reward_threshold: float = -5.0  # Threshold to switch from RL to MPC
+    window_size: int = 50  # Window size for computing average reward
+
+
+class RL2MPCCallback(TrainingProgressCallback):
+    """Callback that stops training when reward threshold is reached."""
+
+    def __init__(
+        self,
+        reward_threshold: float,
+        window_size: int = 50,
+        check_freq: int = 100,
+        verbose: int = 1,
+    ):
+        super().__init__(check_freq=check_freq, verbose=verbose)
+        self.reward_threshold = reward_threshold
+        self.window_size = window_size
+        self.threshold_reached = False
+
+    def _on_step(self) -> bool:
+        should_continue = super()._on_step()
+
+        # Check threshold on episode completion
+        if len(self.episode_rewards) >= self.window_size:
+            recent_mean = np.mean(self.episode_rewards[-self.window_size :])
+            if recent_mean >= self.reward_threshold:
+                print(f"\nReward threshold {self.reward_threshold} reached!")
+                print(f"Recent mean reward: {recent_mean:.2f}")
+                self.threshold_reached = True
+                return False
+        return should_continue
+
+
+class RL2MPCPolicy(Policy[ObsType, ActType]):
+    """Policy that combines RL and MPC."""
+
+    def __init__(self, seed: int, config: RL2MPCConfig | None = None) -> None:
+        """Initialize policy."""
+        super().__init__(seed)
+        self.config = config or RL2MPCConfig()
+
+        # Create component policies
+        self.rl_policy: RLPolicy[ObsType, ActType] = RLPolicy(
+            seed, self.config.rl_config
+        )
+        self.mpc_policy: MPCPolicy[ObsType, ActType] = MPCPolicy(
+            seed, self.config.mpc_config
+        )
+
+        # State tracking
+        self._threshold_reached = False
+        self._start_state: ObsType | None = None
+        self._using_mpc = False
+        self._env: gym.Env | None = None
+
+    @property
+    def requires_training(self) -> bool:
+        """Whether this policy requires training data and training."""
+        return True
+
+    def initialize(self, env: gym.Env) -> None:
+        """Initialize both RL and MPC components."""
+        self._env = env
+        self.rl_policy.initialize(env)
+        self.mpc_policy.initialize(env)
+
+    def configure_context(self, context: PolicyContext) -> None:
+        """Configure policy context."""
+        self.mpc_policy.configure_context(context)
+
+    def train(self, env: gym.Env, train_data: TrainingData) -> None:
+        """Train RL policy until threshold."""
+        if not hasattr(self, "rl_policy") or not hasattr(self, "mpc_policy"):
```

Comment:
hmm wouldn't this never trigger? aren't `self.rl_policy` and `self.mpc_policy` initialized in `__init__`?